File: app/src/main/java/ch/bailu/aat/services/sensor/bluetooth_le/Revolution.java
package ch.bailu.aat.services.sensor.bluetooth_le;

public class Revolution {

    private final Rollover time = new Rollover();
    private final Rollover revolution = new Rollover();


    public void addUINT32(int t, long r) {
        time.add(t);
        revolution.addUINT32(r);
    }

    public void add(int t, int r) {
        time.add(t);
        revolution.add(r);
    }


    public int rpm() {
        int rpm = 0;
        final int time_for_one_revolution = intervall();

        if (time_for_one_revolution > 0)
            rpm = ID.MINUTE / time_for_one_revolution;

        return rpm;
    }


    public int intervall() {
        if (revolution.getDelta() > 0)
            return time.getDelta() / revolution.getDelta();

        return 0;
    }

    public float getSpeedSI(float circumferenceSI) {
        if (circumferenceSI > 0) {
            final int time1024 = intervall();

            if (time1024 > 0) {
                return (circumferenceSI * 1024f) / time1024;
            }
        }

        return 0f;
    }

    public long getTotalRevolutions() {
        return revolution.getTotal();
    }

    public boolean isInitialized() {
        return time.isInitialized() && revolution.isInitialized();
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/services/sensor/bluetooth_le/Revolution.kt
package ch.bailu.aat.services.sensor.bluetooth_le

class Revolution {
    private val time = Rollover()
    private val revolution = Rollover()
    fun addUINT32(t: Int, r: Long) {
        time.add(t.toLong())
        revolution.addUINT32(r)
    }

    fun add(t: Int, r: Int) {
        time.add(t.toLong())
        revolution.add(r.toLong())
    }

    fun rpm(): Int {
        var rpm = 0
        val timeForOneRevolution = interval()
        if (timeForOneRevolution > 0) rpm = ID.MINUTE / timeForOneRevolution
        return rpm
    }

    private fun interval(): Int {
        return if (revolution.delta > 0) time.delta / revolution.delta else 0
    }

    fun getSpeedSI(circumferenceSI: Float): Float {
        if (circumferenceSI > 0) {
            val time1024 = interval()
            if (time1024 > 0) {
                return circumferenceSI * 1024f / time1024
            }
        }
        return 0f
    }

    val totalRevolutions: Long
        get() = revolution.total
    val isInitialized: Boolean
        get() = time.isInitialized && revolution.isInitialized
}
